src/saveload/ser.rs

```diff
@@ -0,0 +1,166 @@
+/// For serializing without such recursion see `serialize` function.
+pub fn serialize_recursive<'a, M, E, T, S>(
+    entities: &Entities<'a>,
+    storages: &<T as Storages<'a>>::ReadStorages,
+    markers: &mut WriteStorage<'a, M>,
+    allocator: &mut FetchMut<'a, M::Allocator>,
+    serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+    M: Marker,
+    E: Display,
+    T: Components<M::Identifier, E>,
+    S: Serializer,
+{
+    let mut serseq = serializer.serialize_seq(None)?;
+    let mut to_serialize = (&**entities, &*markers)
+        .join()
+        .map(|(e, m)| (e, *m))
+        .collect::<Vec<_>>();
+    while !to_serialize.is_empty() {
+        let mut add = vec![];
+        {
+            let mut ids = |entity| -> Option<M::Identifier> {
+                match markers.get(entity).cloned() {
+                    Some(marker) => Some(marker.id()),
+                    None => {
+                        let marker = allocator.mark(entity, markers);
+                        add.push((entity, marker));
+                        Some(marker.id())
+                    }
+                }
+            };
+            for (entity, marker) in to_serialize {
+                serseq.serialize_element(&EntityData::<M, E, T> {
+                    marker,
+                    components: T::save(entity, storages, &mut ids).map_err(ser::Error::custom)?,
+                })?;
+            }
+        }
+        to_serialize = add;
+    }
+    serseq.end()
+}
+
+
+/// Serialize components from specified storages via `SerializableComponent::save`
+/// of all marked entities with provided serializer.
+/// When the component gets serialized with `SerializableComponent::save` method
+/// closure passed in `ids` arguemnt returns `None` for unmarked `Entity`.
+/// In this case `SerializableComponent::save` may perform workaround (forget about `Entity`)
+/// or fail.
+/// So function doesn't recursively mark referenced entities.
+/// For recursive marking see `serialize_recursive`
+pub fn serialize<'a, M, E, T, S>(
+    entities: &Entities<'a>,
+    storages: &<T as Storages<'a>>::ReadStorages,
+    markers: &ReadStorage<'a, M>,
+    serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+    M: Marker,
+    E: Display,
+    T: Components<M::Identifier, E>,
+    S: Serializer,
+{
+    let mut serseq = serializer.serialize_seq(None)?;
+    let ids = |entity| -> Option<M::Identifier> { markers.get(entity).map(Marker::id) };
+    for (entity, marker) in (&**entities, &*markers).join() {
+        serseq.serialize_element(&EntityData::<M, E, T> {
+            marker: *marker,
+            components: T::save(entity, storages, &ids).map_err(ser::Error::custom)?,
+        })?;
+    }
+    serseq.end()
+}
+
+/// This type implements `Serialize` so that it may be used in generic environment
+/// where `Serialize` implementer is expected.
+/// It may be constructed manually with `WorldSerialize::new`.
+/// Or fetched by `System` as `SystemData`.
```

Comment:
s/by/from/

Reply:
I'm not sure here. It's not fetched from `System` as it isn't in `System`. `System` fetches `SystemData` from `World`.